# Gemika’s Enchanted Guide to Iris Dataset with Magic and Machine Learning 🌟🧙‍♂️ (Part #2)

![Chapter 02](images/chapter-02.jpg "Chapter 02")

Welcome back, my brave young witches and wizards, to the hallowed halls of Hogwarts’ data dungeons! I’m Professor Gerry Leo Nugroho, your guide, and today we’re casting **Exploratory Data Analysis (EDA)** on the Iris Dataset. Think of it as shouting *Lumos* to reveal the hidden secrets of these enchanted flowers—unveiling petal and sepal mysteries like a treasure chest in the Room of Requirement! My 9-year-old Gryffindor, Gemika, peeks over my shoulder, asking, “_Are we spying on the flowers like Hermione in the Restricted Section_?” Exactly! EDA is our sneaky peek into the dataset’s soul, counting blooms, measuring quirks, and spotting patterns faster than Harry dodging a [Bludger](https://harrypotter.fandom.com/wiki/Bludger).

---

## Chapter 2. Waving the Wand of Exploration: Sneaky Peeks with EDA 🪄

Gemika Haziq Nugroho, my 9-year-old Gryffindor whirlwind, peers over my shoulder as we unfurl this magic. “_Abi, are we spying on the flowers like Hermione sneaking into the Restricted Section_?” he asks, eyes wide as a Niffler spotting gold. Precisely, my little lion cub! **EDA** is our sneaky peek into the dataset’s soul—counting blooms, measuring their quirks. It’s the first step to mastering the wizardry of numbers!  

This isn’t just about flowers—it’s about learning to ask questions like a true Hogwarts hero. Are `Setosa`’s petals shorter than `Virginica`’s? Do `Versicolor`’s sepals wiggle in the middle? With **EDA**, we’re not just looking; we’re seeing, like using the Marauder’s Map to spot mischief. So, grab your wand, Gemika, and let’s explore this enchanted garden together!

---

### 2.1 The Code & Algorithm

Time to cast our EDA spells with Python—our magical quill scribbling on the parchment of Jupyter Lab! Here’s the incantation to peek into the Iris Dataset:

In [9]:
# Import pandas library for data handling
import pandas as pd  

# Load the Iris dataset from an online source
try:
    iris = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')
    print("✅ Dataset successfully loaded!")
except Exception as e:
    print("❌ Error loading dataset:", e)
    exit()  # Exit the script if data loading fails

# Display the first 5 rows of the dataset
print("\n🌸 First 5 blooms in the Iris dataset:")
print(iris.head())  # Shows the first few entries

# Provide a statistical summary of numerical columns
print("\n🔢 Magical summary of numerical features:")
print(iris.describe())  # Summary statistics (count, mean, std, min, max, etc.)

# Check for missing values in the dataset
print("\n🔍 Checking for missing values in the dataset:")
missing_values = iris.isnull().sum()
print(missing_values)

# Display dataset information, including data types and memory usage
print("\n📜 Dataset Information:")
print(iris.info())

# Show unique species available in the dataset
print("\n🦉 Unique Iris species in the dataset:")
print(iris['species'].unique())  # Displays unique species names  

✅ Dataset successfully loaded!

🌸 First 5 blooms in the Iris dataset:
   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

🔢 Magical summary of numerical features:
       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     

In [10]:
# Summon our owl, pandas, and the Iris scroll  
import pandas as pd  
iris = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')  

# Cast Lumos with head()—peek at the first 5 flowers  
print("First 5 blooms:")  
print(iris.head())  

# Wave the wand with describe()—reveal the numbers’ secrets  
print("\nMagical summary:")  
print(iris.describe())

First 5 blooms:
   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

Magical summary:
       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


Run this, and—*poof!*—the screen glows with five flower rows and a summary of their traits: counts, averages, and ranges, like a Hogwarts report card!  

Gemika gasps, “_It’s like the flowers are talking!_” Indeed, my young wizard—`head()` shows us a glimpse, while `describe()` spills the beans: how many flowers `(150!)`, their smallest petal `(1.0!)`, their biggest sepal `(7.9!)`. It’s like asking Professor Flitwick to summarize a charm—quick, clear, and dazzling. This is our *Revelio* spell, uncovering the dataset’s heartbeat!  

Think of this as potion prep: before brewing, we check our ingredients. EDA with `head()` and `describe()` is our first stir of the cauldron, simple yet powerful—like levitating a feather with _Wingardium Leviosa_. Soon, we’ll add fancier spells, but for now, let’s marvel at these numbers dancing before us! 🧙‍♂️📊

---

### 2.2 Hogwarts Application

Imagine Hagrid lumbering into the Great Hall, his beard full of twigs, pleading, “_Gerry, I need help sortin’ me magical creatures for class_!” With EDA, we can peek at their traits just like we do with Iris flowers. Are [Hippogriffs](https://harrypotter.fandom.com/wiki/Hippogriff) taller than [Thestrals](https://harrypotter.fandom.com/wiki/Thestral)? Do Bowtruckles have shorter claws than Nifflers? Our `head()` spell spots the first few, and `describe()` tells us their averages—perfect for Hagrid’s lesson plans!  

Gemika bounces, “Can we spy on dragons too, Dad?” Oh, yes! Picture us sneaking into the Triwizard Tournament, using EDA to measure dragon wingspans or fire-breath range. It’s not just creatures—EDA could count Chocolate Frogs in Honeydukes or track broomstick speeds for Quidditch tryouts. Wherever there’s mystery at Hogwarts, our sneaky peek spell saves the day!  

This is wizardry for the curious—like Fred and George plotting pranks. By exploring data, we help Hagrid, cheer Harry at Quidditch, or even sneak a peek at Snape’s potion stash (don’t tell him!). EDA turns numbers into stories, and at Hogwarts, every story’s a spell waiting to be cast! 🐉🌟  

---

### 2.3 Quiz from Gemika Haziq Nugroho  

Gemika twirls his wand, grinning like a Cheshire Cat. “Abi, how do we know what the numbers are whispering? Are they secrets from the flowers?”  

- **A)** They’re just boring numbers with no meaning, like Peeves’ bad jokes.  
- **B)** They tell us about the flowers’ sizes and shapes, like clues in a riddle!  
- **C)** They’re passwords to the Slytherin common room—shh, don’t tell Draco!  

What’s your answer, clever witch or wizard? Hint: Think of what *Lumos* shows us in the dark! 🦁🪄

---

### 2.4 Next Chapter: A Cauldron of Colors Awaits!

Hold tight to your broomsticks, because next we’re brewing a potion of interactive visualizations! We’ll stir up dazzling charts with `seaborn` and `plotly`—think moving portraits or a Quidditch pitch map! It’ll be so grand, even Draco Malfoy might sneak a peek. Get ready for colors, giggles, and maybe a dragon-sized surprise!